google-cloud-pubsub/src/main/java/com/google/cloud/pubsub/v1/MessageDispatcher2.java

```diff
@@ -0,0 +1,284 @@
+                List<String> extendIds = new ArrayList<>(extensionSet.keySet());
+                for (List<String> batch : Lists.partition(extendIds, MAX_CHANGE_PER_REQUEST)) {
+                  sendRequest(
+                      StreamingPullRequest.newBuilder()
+                          .addAllModifyDeadlineAckIds(batch)
+                          .addAllModifyDeadlineSeconds(
+                              Collections.nCopies(batch.size(), DEADLINE_EXTENSION_SEC))
+                          .build());
+                }
+              }
+            },
+            KEEPALIVE_SEC,
+            KEEPALIVE_SEC,
+            TimeUnit.SECONDS);
+  }
+
+  synchronized void shutdown() {
+    ackNackJob.cancel(false);
+    extensionJob.cancel(false);
+  }
+
+  private boolean drainAckNacks(StreamingPullRequest.Builder builder) {
+    while (builder.getAckIdsCount() < MAX_CHANGE_PER_REQUEST) {
+      String id = idsToAck.poll();
+      if (id == null) {
+        break;
+      }
+      builder.addAckIds(id);
+    }
+
+    while (builder.getModifyDeadlineAckIdsCount() < MAX_CHANGE_PER_REQUEST) {
+      ModAckItem modAck = modAcks.poll();
+      if (modAck == null) {
+        break;
+      }
+      builder.addModifyDeadlineAckIds(modAck.ackId());
+      builder.addModifyDeadlineSeconds(modAck.seconds());
+    }
+
+    return builder.getAckIdsCount() == MAX_CHANGE_PER_REQUEST
+        || builder.getModifyDeadlineAckIdsCount() == MAX_CHANGE_PER_REQUEST;
+  }
+
+  void processReceivedMessages(List<ReceivedMessage> messages, Runnable doneCallback) {
+    if (messages.isEmpty()) {
+      doneCallback.run();
+      return;
+    }
+
+    for (ReceivedMessage message : messages) {
+      modAcks.add(ModAckItem.create(message.getAckId(), DEADLINE_EXTENSION_SEC));
+    }
+
+    AtomicInteger remaining = new AtomicInteger(messages.size());
+    for (ReceivedMessage message : messages) {
+      workQueue.add(new WorkItem(message, remaining, doneCallback));
+      extensionSet.put(message.getAckId(), true);
+    }
+    workMessages();
+  }
+
+  private synchronized void workMessages() {
+    for (; ; ) {
+      final WorkItem item = workQueue.peek();
+      if (item == null) {
+        return;
+      }
+
+      try {
+        flowController.reserve(1, item.message.getMessage().getSerializedSize());
+      } catch (FlowControlException e) {
+        return;
+      }
+      workQueue.poll();
+
+      final Consumer consumer = new Consumer(item);
+      executor.submit(
+          new Runnable() {
+            @Override
+            public void run() {
+              try {
+                receiver.receiveMessage(item.message.getMessage(), consumer);
+              } catch (Exception e) {
+                consumer.throwException(e);
+              }
+            }
+          });
+    }
+  }
+
+  private void sendRequest(StreamingPullRequest request) {
+    Connection connection = null;
+    try {
+      connection = connections.take();
```

Comment:
This logic is confusing, it's important we send acks and mod acks through the same connections that received the messages.

Reply:
I thought I was being smart round-robin-ing the connections :(

Could you provide some context here? The current logic seems to work fine on both desktop and loadtest. Might it not work in the future?